modules/33acrossAnalyticsAdapter.js

```diff
@@ -0,0 +1,475 @@
+      this.complete(transactionId);
+    };
+
+    log.info('All remaining transactions flushed.' + (reason ? ` Reason: ${reason}` : ''));
+  }
+
+  #flushTransactions() {
+    this.#clearSendTimeout();
+    this.#transactionsPending = new Set();
+    this.#onComplete();
+  }
+
+  // gulp-eslint is using eslint 6, a version that doesn't support private method syntax
+  // eslint-disable-next-line no-dupe-class-members
+  #clearSendTimeout() {
+    return clearTimeout(this.#sendTimeoutId);
+  }
+
+  // eslint-disable-next-line no-dupe-class-members
+  #restartSendTimeout() {
+    this.#clearSendTimeout();
+
+    this.#sendTimeoutId = setTimeout(() => {
+      if (this.#sendTimeout !== 0) {
+        log.warn(`Timed out waiting for ad transactions to complete. Sending report.`);
+      }
+
+      this.#flushTransactions();
+    }, this.#sendTimeout);
+  }
+}
+
+/**
+ * initialized during `enableAnalytics`
+ */
+const locals = {
+  /** @type {Object<string, TransactionManager>} - one manager per auction */
+  transactionManagers: {},
+  /** @type {AnalyticsCache=} */
+  cache: undefined,
+  /** sets all locals to undefined */
+  reset() {
+    this.transactionManagers = {};
+    this.cache = undefined;
+  }
+}
+
+/**
+ * @typedef {Object} AnalyticsAdapter
+ * @property {function} track
+ * @property {function} enableAnalytics
+ * @property {function} disableAnalytics
+ * @property {function} [originEnableAnalytics]
+ * @property {function} [originDisableAnalytics]
+ * @property {function} [_oldEnable]
+ */
+
+/**
+ * @type {AnalyticsAdapter}
+ */
+const analyticsAdapter = Object.assign(
+  buildAdapter({ analyticsType: 'endpoint' }),
+  { track: analyticEventHandler }
+);
+
+analyticsAdapter.originEnableAnalytics = analyticsAdapter.enableAnalytics;
+analyticsAdapter.enableAnalytics = enableAnalyticsWrapper;
+
+/**
+ * @typedef {Object} AnalyticsConfig
+ * @property {string} provider - set by pbjs at module registration time
+ * @property {Object} options
+ * @property {string} options.pid - Publisher/Partner ID
+ * @property {string} [options.endpoint=DEFAULT_ENDPOINT] - Endpoint to send analytics data
+ * @property {number} [options.timeout=DEFAULT_TRANSACTION_TIMEOUT] - Timeout for sending analytics data
+ */
+
+/**
+ * @param {AnalyticsConfig} config Analytics module configuration
+ */
+function enableAnalyticsWrapper(config) {
+  const { options } = config;
+
+  const pid = options.pid;
+  if (!pid) {
+    log.error('No partnerId provided for "options.pid". No analytics will be sent.');
+
+    return;
+  }
+
+  const endpoint = calculateEndpoint(options.endpoint);
+  this.getUrl = () => endpoint;
+
+  const timeout = calculateTransactionTimeout(options.timeout);
+  this.getTimeout = () => timeout;
+
+  locals.cache = {
+    pid,
+    auctions: {},
+  };
+
+  analyticsAdapter.originEnableAnalytics(config);
+}
+
+/**
+ * @param {string} [endpoint]
+ * @returns {string}
+ */
+function calculateEndpoint(endpoint) {
```

Comment:
Since we currently don't expect that publishers will specify the `endpoint` parameter (and we might not even mention it in our external documentation), the norm will actually be *not* specifying an `endpoint` at all. So I'd limit logging that `Invalid endpoint...` message to a case where the the publisher does specify an `endpoint` but it's not valid. One way to do this might be to change the method signature to `calculateEndpoint(endpoint = DEFAULT_ENDPOINT)`.

Reply:
This was a misconception on my part about how we wanted to enable this for publishers. But this makes sense; I'll make the appropriate changes here and in docs.